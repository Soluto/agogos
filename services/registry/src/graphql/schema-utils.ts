import {
    DefinitionNode,
    DocumentNode,
    FieldDefinitionNode,
    InputValueDefinitionNode
} from "graphql";
import {
    clone,
    equals,
    flatten,
    groupBy,
    indexBy,
    map,
    pipe,
    unionWith,
    uniq,
    values
} from "ramda";

function validateSameFieldDefs(
    fieldDefA: FieldDefinitionNode | InputValueDefinitionNode,
    fieldDefB: FieldDefinitionNode | InputValueDefinitionNode
) {
    const { directives: _, ...fieldDefWithNoDirectivesA } = fieldDefA;
    const { directives: __, ...fieldDefWithNoDirectivesB } = fieldDefB;
    if (!equals(fieldDefWithNoDirectivesA, fieldDefWithNoDirectivesB)) {
        throw new Error(`Same field definition should be identical except for directives - field ${
            fieldDefA.name.value
            }`);
    }
}

function mergeSameFieldDefinitions(
    fieldDefinitions: ReadonlyArray<
        FieldDefinitionNode | InputValueDefinitionNode
    >
): ReadonlyArray<FieldDefinitionNode | InputValueDefinitionNode> {
    if (fieldDefinitions.length === 0) { return []; }
    if (fieldDefinitions.length === 1) { return fieldDefinitions; }

    const [fieldDefA, fieldDefB, ...rest] = fieldDefinitions;
    validateSameFieldDefs(fieldDefA, fieldDefB);

    const { directives: _, ...otherFields } = fieldDefA;
    const fieldWithMergedDirectives = {
        ...otherFields,
        directives: uniq([
            ...(fieldDefA.directives || []),
            ...(fieldDefB.directives || [])
        ])
    };

    return mergeSameFieldDefinitions([fieldWithMergedDirectives, ...rest]);
}

const mergeFields = (
    l: ReadonlyArray<FieldDefinitionNode | InputValueDefinitionNode>,
    r: ReadonlyArray<FieldDefinitionNode | InputValueDefinitionNode>
) =>
    pipe(
        groupBy(
            (fieldDef: FieldDefinitionNode | InputValueDefinitionNode) =>
                fieldDef.name.value
        ),
        values,
        map(mergeSameFieldDefinitions),
        flatten
    )([...l, ...r]);

export function mergeDocuments(schemas: DocumentNode[]) {
    return clone(schemas).reduce((acc: DocumentNode, next: DocumentNode) => {
        const index = indexBy((x: DefinitionNode) => {
            if (
                x.kind === "ObjectTypeDefinition" ||
                x.kind === "UnionTypeDefinition" ||
                x.kind === "InputObjectTypeDefinition"
            ) {
                return `${x.kind}_${x.name.value}`;
            }
            return "";
        })(acc.definitions);

        const get = (kind: string, value: string) => index[`${kind}_${value}`];

        next.definitions.forEach(x => {
            if (
                (x.kind !== "ObjectTypeDefinition" &&
                    x.kind !== "UnionTypeDefinition" &&
                    x.kind !== "InputObjectTypeDefinition") ||
                !get(x.kind, x.name.value)
            ) {
                (acc.definitions as any).push(x);
                return;
            }

            const ref = get(x.kind, x.name.value);
            if (
                x.kind === "ObjectTypeDefinition" ||
                x.kind === "InputObjectTypeDefinition"
            ) {
                (ref as any).directives = uniq([
                    ...(ref as any).directives,
                    ...(x.directives || [])
                ]);
                (ref as any).fields = mergeFields((ref as any).fields, x.fields || []);

                if (x.kind === "ObjectTypeDefinition") {
                    (ref as any).interfaces = uniq([
                        ...(ref as any).interfaces,
                        ...(x.interfaces || [])
                    ]);
                }
                return;
            }

            if (x.kind === "UnionTypeDefinition") {
                (ref as any).types = unionWith(equals, (ref as any).types, x.types || []);
                return;
            }
        });

        return acc;
    });
}
