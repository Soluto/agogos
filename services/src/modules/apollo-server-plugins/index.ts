import { PluginDefinition } from 'apollo-server-core';
import { transformApolloServerPlugins as applyPluginsForApolloServerPlugins } from '../plugins';
import { createBasicPolicyPlugin } from './base-policy';
import { createLoggingPlugin } from './logging';
import { createMetricsPlugin } from './metrics';

export default function getPlugins(): PluginDefinition[] {
  const baseApolloServerPlugins = [createBasicPolicyPlugin, createLoggingPlugin, createMetricsPlugin];
  const plugins = applyPluginsForApolloServerPlugins(baseApolloServerPlugins);
  return plugins;
}
