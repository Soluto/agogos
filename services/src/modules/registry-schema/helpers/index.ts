import proceedPolicy from './policy-helper';

export { proceedPolicy };
