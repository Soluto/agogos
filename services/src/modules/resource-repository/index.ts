export * from './types';

export { ResourceRepository } from './resource-repository';
export { CompositeResourceRepository } from './composite';
export { S3ResourceRepository } from './s3-repository';
export { FileSystemResourceRepository } from './fs-repository';
export { pollForUpdates } from './stream';
export { applyResourceGroupUpdates, applyResourceGroupDeletions } from './actions';
export { default as getResourceRepository } from './get-resource-repository';
