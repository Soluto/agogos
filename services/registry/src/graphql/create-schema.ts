import gql from "graphql-tag";
import { mergeDocuments } from "./schema-utils";
import { DocumentNode } from "graphql";

type GqlSource = {
  source: string;
  gql: string;
};

export type GqlSources = { [source: string]: string };

function parseGqlSource(gqlSource: GqlSource) {
  try {
    return gql(gqlSource.gql);
  } catch (e) {
    throw {
      message: `Failed to load schema for source -  ${gqlSource.source} - ${
        e.message
      }`,
      schema: gqlSource,
      error: e
    };
  }
}

function mergeAllDocuments(docs: DocumentNode[]) {
  try {
    return mergeDocuments(docs);
  } catch (e) {
    throw {
      message: `Failed to merge schemas - ${e.message}`,
      schemas: docs,
      error: e
    };
  }
}

export const makeGqlDocumentFromGqlSources = (gqlSources: GqlSources) => {
  const documentNodes = Object.entries(gqlSources).map(([source, gql]) =>
    parseGqlSource({ source, gql })
  );
  return mergeAllDocuments(documentNodes);
};
