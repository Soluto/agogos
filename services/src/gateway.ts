import * as fastify from 'fastify';
import * as fastifyMetrics from 'fastify-metrics';

import * as config from './modules/config';
import logger from './modules/logger';
import {handleSignals, handleUncaughtErrors} from './modules/shutdownHandler';
import {createStitchGateway} from './modules/gateway';
import {pollForUpdates, S3ResourceRepository} from './modules/resource-repository';

async function run() {
    logger.info('Stitch gateway booting up...');

    const resourceRepository = new S3ResourceRepository();

    const {server, dispose} = createStitchGateway({
        resourceGroups: pollForUpdates(resourceRepository, config.resourceUpdateInterval),
        tracing: config.enableGraphQLTracing,
        playground: config.enableGraphQLPlayground,
        introspection: config.enableGraphQLIntrospection,
    });

    const app = fastify();
    app.register(fastifyMetrics, {endpoint: '/metrics'});
    app.register(server.createHandler({path: '/graphql'}));
    const address = await app.listen(config.httpPort, '0.0.0.0');
    logger.info({address}, 'Stitch gateway started');

    handleSignals(dispose);
    handleUncaughtErrors();
}

// Only run when file is being executed, not when being imported
if (require.main === module) {
    run();
}
