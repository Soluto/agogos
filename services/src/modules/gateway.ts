import * as fastify from 'fastify';
import { ApolloServer, Config } from 'apollo-server-fastify';
import { Observable } from 'rxjs';
import { createGraphQLService } from './graphql-service';
import { RESTDirectiveDataSource } from './directives/rest';
import { ResourceGroup } from './resource-repository';
import { ExportTrackingExtension } from './exports';
import {
  createBasicPolicyPlugin,
  createLoggingPlugin,
  createMetricsPlugin,
  createArgumentInjectionPlugin,
} from './plugins';

export interface GatewayConfig extends Config {
  resourceGroups: Observable<ResourceGroup>;
  fastifyInstance: Pick<fastify.FastifyInstance, 'metrics'>;
}

export function createStitchGateway(config: GatewayConfig) {
  const { resourceGroups, fastifyInstance, ...apolloConfig } = config;
  const gateway = createGraphQLService({ resourceGroups });
  const server = new ApolloServer({
    ...apolloConfig,
    gateway,
    extensions: [() => new ExportTrackingExtension()],
    plugins: [
      createBasicPolicyPlugin,
      createLoggingPlugin(),
      createMetricsPlugin(fastifyInstance),
      createArgumentInjectionPlugin(),
    ],
    subscriptions: false,
    context(request: fastify.FastifyRequest) {
      const ctx = {
        request,
      };
      return ctx;
    },
    dataSources() {
      return {
        rest: new RESTDirectiveDataSource(),
      };
    },
  });

  return {
    server,
    dispose: async () => {
      await server.stop();
      gateway.dispose();
    },
  };
}
