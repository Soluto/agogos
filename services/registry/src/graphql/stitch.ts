import {DocumentNode, parse} from 'graphql';
import {mergeDocuments} from './schema-utils';
import {AggObjsByName} from '../sync/sync-service';
import {SchemaConfig} from '../sync/object-types';

function parseGqlSource(schemaName: string, schema: string) {
    try {
        return parse(schema);
    } catch (e) {
        throw {
            error: e,
            message: `Failed to load schema for source -  ${schemaName} - ${e.message}`,
            schema: {schemaName, schema},
        };
    }
}

function mergeAllDocuments(docs: DocumentNode[]) {
    try {
        return mergeDocuments(docs);
    } catch (e) {
        throw {
            error: e,
            message: `Failed to merge schemas - ${e.message}`,
            schemas: docs,
        };
    }
}

export const stitchSchemas = (gqlSchemas: AggObjsByName) => {
    const documentNodes = Object.entries(gqlSchemas).map(([schemaName, schema]: [string, SchemaConfig]) =>
        parseGqlSource(schemaName, schema.definition)
    );
    return mergeAllDocuments(documentNodes);
};
