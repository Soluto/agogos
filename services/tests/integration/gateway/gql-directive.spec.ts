import { createTestClient, ApolloServerTestClient } from 'apollo-server-testing';
import * as Rx from 'rxjs';
import * as nock from 'nock';
import { makeExecutableSchema } from 'graphql-tools';
import { gql } from 'apollo-server-core';
import { graphqlSync, GraphQLSchema, print } from 'graphql';
import { createStitchGateway } from '../../../src/modules/gateway';
import { beforeEachDispose } from '../before-each-dispose';
import { sleep } from '../../helpers/utility';
import { Schema, Upstream } from '../../../src/modules/resource-repository';

const miriam = {
  employeeId: '1',
  name: 'miriam',
  age: 97,
};
const ethel = {
  employeeId: '2',
  name: 'ethel',
  age: 94,
  bestFriend: miriam,
};
const employees: { [id: string]: any } = {
  '1': miriam,
  '2': ethel,
};

const remoteSchema = makeExecutableSchema({
  typeDefs: gql`
    type Employee {
      employeeId: ID!
      name: String!
      age: Int!
      bestFriend: Employee
    }

    type Query {
      employee(id: ID!): Employee
    }
  `,
  resolvers: {
    Query: {
      employee: (_, { id }: { id: string }) => employees[id],
    },
  },
});

const remoteHost = 'http://localhost:1111';
const remoteLogicalHost = 'https://my-service';

const schema: Schema = {
  metadata: {
    namespace: 'namespace',
    name: 'name',
  },
  schema: print(gql`
    type Employee {
      employeeId: ID!
      name: String!
      age: Int!
      extraFieldThatsNotOnTheRemoteGql: String! @localResolver(value: "surprise!")
      bestFriend: Employee
    }

    type Query {
      employee(employeeId: ID!): Employee
        @gql(
          url: "${remoteHost}/graphql"
          fieldName: "employee"
          arguments: { id: "{args.employeeId}" }
          timeoutMs: 100
        )
    }
  `),
};

const upstream: Upstream = {
  metadata: {
    namespace: 'namespace',
    name: 'name',
  },
  host: new URL(remoteLogicalHost).host,
  origin: remoteHost,
};

const resourceGroup = {
  etag: 'etag',
  schemas: [schema],
  upstreams: [upstream],
  upstreamClientCredentials: [],
  policies: [],
};

interface TestCase {
  statusCode?: number;
  delay?: number;
  host?: string;
}

const testCases: [string, TestCase][] = [
  ['Success', {}],
  ['Error 401', { statusCode: 401 }],
  ['Error 500', { statusCode: 500 }],
  ['Timeout', { delay: 2000 }],
  ['Upstream Host', { host: remoteLogicalHost }],
];

describe.each(testCases)('GQL Directive', (testCaseName, { statusCode, delay, host }) => {
  let client: ApolloServerTestClient;

  beforeEachDispose(async () => {
    mockGqlBackend(host ?? remoteHost, remoteSchema, statusCode, delay);

    const stitch = createStitchGateway({
      resourceGroups: Rx.of(resourceGroup),
      fastifyInstance: { metrics: undefined as any },
    });
    client = createTestClient(stitch.server);

    // Wait for introspection queries
    await sleep(300);

    return () => {
      nock.cleanAll();
      return stitch.dispose();
    };
  });

  it(`Remote GraphQL server call - ${testCaseName}`, async () => {
    const response = await client
      .query({
        query: gql`
          query EmpById($id: ID!) {
            employee(employeeId: $id) {
              name
              bestFriend {
                extraFieldThatsNotOnTheRemoteGql
                employeeId
                age
              }
            }
          }
        `,
        variables: { id: 2 },
      })
      .catch(e => e.response);

    expect(response).toMatchSnapshot();
  });
});

function mockGqlBackend(host: string, schema: GraphQLSchema, statusCode = 200, delay = 0) {
  nock(host)
    .persist()
    .post('/graphql', body => body.operationName === 'IntrospectionQuery')
    .reply(200, (_, body: any) =>
      graphqlSync({
        schema,
        source: body.query,
        variableValues: body.variables,
        operationName: body.operationName,
      })
    )
    .post('/graphql', body => body.operationName !== 'IntrospectionQuery')
    .delay(delay)
    .reply(statusCode, (_, body: any) =>
      graphqlSync({
        schema,
        source: body.query,
        variableValues: body.variables,
        operationName: body.operationName,
      })
    );
}
