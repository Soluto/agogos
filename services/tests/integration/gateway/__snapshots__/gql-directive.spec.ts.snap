// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GQL Directive Remote GraphQL server call - Error 401 1`] = `
Object {
  "data": Object {
    "employee": null,
  },
  "errors": Array [
    [GraphQLError: Response not successful: Received status code 401],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`GQL Directive Remote GraphQL server call - Error 500 1`] = `
Object {
  "data": Object {
    "employee": null,
  },
  "errors": Array [
    [GraphQLError: Response not successful: Received status code 500],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`GQL Directive Remote GraphQL server call - Success 1`] = `
Object {
  "data": Object {
    "employee": Object {
      "bestFriend": Object {
        "age": 97,
        "employeeId": "1",
        "extraFieldThatsNotOnTheRemoteGql": "surprise!",
      },
      "name": "ethel",
    },
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`GQL Directive Remote GraphQL server call - Timeout 1`] = `
Object {
  "data": Object {
    "employee": null,
  },
  "errors": Array [
    [GraphQLError: network timeout at: http://localhost:1111/graphql],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`GQL Directive Remote GraphQL server call - Upstream Host 1`] = `
Object {
  "data": Object {
    "employee": null,
  },
  "errors": Array [
    [GraphQLError: Failed reaching remote gql server for introspection (url http://localhost:1111/graphql)],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
